File: java/src/com/algos/topics/li10_tree/bt/revised/Bt_L236_LowestCommonAncestorOfABinaryTree.java
package  com.algos.topics.li10_tree.bt.revised;

import com.algos.topics.models.TreeNode;


public class Bt_L236_LowestCommonAncestorOfABinaryTree {
    public static void main(String[] args) {
        Bt_L236_LowestCommonAncestorOfABinaryTree obj = new Bt_L236_LowestCommonAncestorOfABinaryTree();

        TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

        System.err.println(obj.lowestCommonAncestor(root, root.left.left, root.left.right.left));
    }

    // Problem: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/
    // Idea: Find num1 and num2. If found return its root else return null.
    // At LCA root, you will recieve not null responses from its left and right. Return that root as LCA.
    // Time: O(n)
    // space: O(1)
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root.val == p.val || root.val == q.val)
			return root;

		TreeNode left = lowestCommonAncestor(root.left, p, q);
		TreeNode right = lowestCommonAncestor(root.right, p, q);
		if(left == null) {
			return right;
		}else if(right == null){
			return left;
		}else {
			return root;
		}
    }
}

File: dsa/patterns/10_trees/bt/python/bt_l236__lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/
# Idea: Recursively search for p and q in left and right subtrees.
# If root matches p or q, return root.
# LCA is where left and right recursive calls both return non-null.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def lowest_common_ancestor(root, p, q):
    if not root or root.val == p.val or root.val == q.val:
        return root

    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)

    if not left:
        return right
    if not right:
        return left

    return root


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

lca = lowest_common_ancestor(root, root.left.left, root.left.right.left)
print(lca.val if lca else None)  # Expected: 3
